resolver/resolver_test.go

```diff
@@ -417,27 +417,68 @@ func TestMultiError(t *testing.T) {
 	}
 }
 
-func TestTruncate(t *testing.T) {
-	tm := *newTruncated()
-	if !tm.Truncated {
+func TestTruncateSetTruncateBit(t *testing.T) {
+	testTruncateSetTruncateBit(t, dns.MinMsgSize)
+}
+
+func TestTruncateEdns0SetTruncateBit(t *testing.T) {
+	testTruncateSetTruncateBit(t, 4096)
+}
+
+func TestTruncateNoSetTruncateBit(t *testing.T) {
+	testTruncateNoSetTruncateBit(t, dns.MinMsgSize)
+}
+
+func TestTruncateEdns0NoSetTruncateBit(t *testing.T) {
+	testTruncateNoSetTruncateBit(t, 4096)
+}
+
+func testTruncateSetTruncateBit(t *testing.T, maxSize uint16) {
+	orig := *newMessage(maxSize)
```

Comment:
The point of `orig` and `tmp` is exactly to handle your previous concern regarding mutating the input. Notice that we dereference (`*truncate()`) where we use it.

For instance:

```
orig := *newMessage(maxSize)
// orig is a concrete value
tmp := orig
// tmp is also a concrete value, it is not orig, although tmp.Answer shares a
// backing array with orig.Answer which could cause a problem if this array
// got sorted later on, for instance.
msg := *truncate(&tmp, true, true)
// tmp is mutated, but never used again (hence it is called 'tmp')
// msg is a new concrete value.
// orig is left unchanged.
...
```